lms/djangoapps/grades/new/subsection_grade.py

```diff
@@ -46,34 +47,101 @@ def compute(self, student, course_structure, scores_client, submissions_scores):
+            for location, (score, weight) in self.locations_to_scores.iteritems()
+        ]
+        PersistentSubsectionGrade.save_grade(
+            user_id=student.id,
+            usage_key=self.location,
+            course_version=getattr(course, 'course_version', ""),
+            subtree_edited_date=subsection.subtree_edited_on,
+            earned_all=self.all_total.earned,
+            possible_all=self.all_total.possible,
+            earned_graded=self.graded_total.earned,
+            possible_graded=self.graded_total.possible,
+            visible_blocks=visible_blocks,
+        )
+
+    def load_from_data(self, model, course_structure, scores_client, submissions_scores):
+        """
+        Load the subsection grade from the persisted model.
+        """
+        for block in model.visible_blocks.blocks:
+            persisted_values = {'weight': block.weight, 'possible': block.max_score}
+            self._compute_block_score(
+                User.objects.get(id=model.user_id),
+                block.locator,
+                course_structure,
+                scores_client,
+                submissions_scores,
+                persisted_values
+            )
+
+        self.graded_total = Score(
+            model.earned_graded,
+            model.possible_graded,
+            True,
+            self.display_name,
+            self.location,
+        )
+        self.all_total = Score(
+            model.earned_all,
+            model.possible_all,
+            False,
+            self.display_name,
+            self.location,
+        )
+
+    def _compute_block_score(
+            self,
+            student,
+            block_key,
+            course_structure,
+            scores_client,
+            submissions_scores,
+            persisted_values=None,
+    ):
+        """
+        Compute score for the given block. If persisted_values is provided, it will be used for possible and weight.
+        """
+        block = course_structure[block_key]
+
+        if getattr(block, 'has_score', False):
             (earned, possible) = get_score(
                 student,
-                descendant,
+                block,
                 scores_client,
                 submissions_scores,
             )
-            if earned is None and possible is None:
-                continue
-
-            # cannot grade a problem with a denominator of 0
-            descendant_graded = descendant.graded if possible > 0 else False
-
-            self.locations_to_scores[descendant.location] = Score(
-                earned,
-                possible,
-                descendant_graded,
-                block_metadata_utils.display_name_with_default_escaped(descendant),
-                descendant.location,
-            )
 
-        self.all_total, self.graded_total = graders.aggregate_scores(
-            self.scores, self.display_name,
-        )
+            weight = block.weight
```

Comment:
Please add a comment here that there's a possible race condition here where the value of `weight` may not be the value when the block was scored - since we are getting the value from 2 different places: from the block_structure and from CSM/Submissions.  This may be more of an issue if the subsection grade is updated asynchronously.